pkg/controller/metrics/provision_underway_collector.go

```diff
@@ -0,0 +1,91 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+var (
+	// ClusterDeployment conditions that could indicate provisioning problems with a cluster
+	provisioningDelayCondition = [...]hivev1.ClusterDeploymentConditionType{
+		hivev1.ProvisionFailedCondition,
+		hivev1.InstallLaunchErrorCondition,
+	}
+)
+
+// provisioning underway metrics collected through a custom prometheus collector
+type provisioningUnderwayCollector struct {
+	client client.Client
+
+	// metricClusterDeploymentProvisionUnderwaySeconds is a prometheus metric for the number of seconds
+	// between when a still provisioning cluster was created and now.
+	metricClusterDeploymentProvisionUnderwaySeconds *prometheus.Desc
+}
+
+// collects the metrics for provisioningUnderwayCollector
+func (cc provisioningUnderwayCollector) Collect(ch chan<- prometheus.Metric) {
+	ccLog := log.WithField("controller", "metrics")
+	ccLog.Info("calculating provisioning underway metrics across all ClusterDeployments")
+
+	// Load all ClusterDeployments so we can accumulate facts about them.
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err := cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+	} else {
+		for _, cd := range clusterDeployments.Items {
+			if cd.DeletionTimestamp == nil {
+				if !cd.Spec.Installed {
```

Comment:
```
if cd.DeletionTimestamp == nil {
  continue
}

if cd.Spec.Installed {
   continue
}

// actually what matters
```

Keeping the actual business logic to the left as much possible makes reading the code more easy, and also helps not-remembering what the context was when reading it, which is necessary when you have it in nested conditionals.

Reply:
These are good advice, up to you if you want to cancel the lgtm and update or let it go @suhanime.

Reply:
Strong +1 to this general advice!